sycl/doc/extensions/proposed/sycl_ext_oneapi_prefetch.asciidoc

```diff
@@ -0,0 +1,437 @@
+cover many real use-cases -- it does not acknowledge the existence of a cache
+hierarchy, and does not provide a mechanism for cooperative (i.e. group)
+prefetches.
+
+This proposal addresses these shortcomings by introducing a set of
+free functions that groups of work-items can use to prefetch data into specific
+levels of cache (controlled via compile-time properties).
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_PREFETCH` to one of the values defined in the table
+below.  Applications can test for the existence of this macro to determine if
+the implementation supports this feature, or applications can test the macro's
+value to determine which of the extension's features the implementation
+supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+
+=== Specifying cache levels
+
+The level of cache targeted by a prefetch is specified using compile-time
+properties. This extension defines hints for four levels of cache,
+corresponding to the four levels of cache currently defined in SYCL (e.g.
+as used by `info::partition_affinity_domain`).
+
+[NOTE]
+====
+Not all devices targeted by SYCL will have four levels of cache. Some of these
+hints may have no meaning on some devices. However, implementations are
+encouraged to map these constants to the closest level of cache available.
+====
+
+When multiple cache levels are specified, the lowest level takes precedence
+(e.g. a request to prefetch into L1 and L4 is treated as a request to prefetch
+into L1). When no cache levels are specified, the default behavior is to
+prefetch into the lowest level cache (i.e. L1).
+
+[NOTE]
+====
+Future hints may alter the default behavior of prefetches with respect to cache
+levels. Any such alterations are expected to be documented in the definition
+of those new hints.
+====
+
+[source,c++]
+----
+namespace sycl::ext::oneapi::experimental {
+
+enum class cache_level {
+  L1,
+  L2,
+  L3,
+  L4,
+};
+
+struct nontemporal;
+
+struct prefetch_hint_key {
+  template <cache_level Level, typename Hint>
+  using value_t = property_value<prefetch_hint_key,
+                                 std::integral_constant<cache_level, Level>,
+                                 Hint>
+};
+
+template <cache_level Level, typename Hint>
+inline constexpr prefetch_hint_key::value_t<Level, Hint> prefetch_hint;
+
+inline constexpr prefetch_hint_key::value_t<cache_level::L1, void> prefetch_hint_L1;
+inline constexpr prefetch_hint_key::value_t<cache_level::L2, void> prefetch_hint_L2;
+inline constexpr prefetch_hint_key::value_t<cache_level::L3, void> prefetch_hint_L3;
+inline constexpr prefetch_hint_key::value_t<cache_level::L4, void> prefetch_hint_L4;
```

Comment:
Since we are adding in some prebuilt hints, should we also add built-ins for prefetch_hint_L1_nontemporal ?  etc  or just leave that to users ?

Reply:
I'm not opposed to adding some additional shorthands.  I shortened to `NT` (e.g. `prefetch_hint_L1_NT`) and added in 18b812a58195ac48a09d58fa57bc4729d3fc439a.